### PR TITLE
ibmcloud: Add end-to-end Terraform configuration

### LINKED ISSUE #41: As an end user, I want a Terraform template that builds and deploys the entire demo environment
Currently there are a set of Terraform templates in confidential-containers/cloud-api-adaptor/ibmcloud that can set up the IBM Cloud VPC environment, build the peer pod Custom Image image and run a demo that uses the peer pod image.

The templates need to be ran in serial by hand, so it would be a better user experience to have the option of running a single Terraform template that runs the existing Terraform templates as (sub)modules, so from and end-user perspective there is one Terraform template that does the VPC resource provisioning, building the peer pod image and running the demo end-to-end, and hence only a single thing to run.

### PR DESCRIPTION
**Pull request by mattarnoatibm**
Add Terraform configuration that does end to end setup and test of the peer pods demo environment

Add documentation for the e2e configuration.

Fixes #41

Signed-off-by: mattarno <mattarno@uk.ibm.com>

### CODE REVIEW

**Comment by mattarnoatibm** [comment]
I’ve refactored the Terraform configurations for the cloud API adaptor so that there’s now a single configuration you can run that sets up and tests the entire demo environment.

I had to do a largeish amount of refactoring on the Terraform configurations that set up the individual steps of the demo to make the end to end configuration possible to run.

In the Terraform configurations that are called as modules by the e2e configuration, for resource parameters that were created earlier by the e2e configuration you need to pass in the ID of the created resource, not the name. This is because if you pass in the name and use the data source to get the ID, the data source will try and get the resource before it has been created. If you’re configuration is actually creating the resource, and hence you know its ID from the output from the module that creates the resource, it makes more sense to pass in the resource ID, which is not known until the resource gets created, to the next module rather than use the resource name and the data source to look up the ID. That enforces the right chain of dependencies so that Terraform executes actions in the right order.

For compatibility with running the Terraform configurations individually, for resource parameters you need to specify using the ID in the e2e configuration there is also the option of specifying those resources by name. The intended use case is when calling those configurations individually, not as part of the e2e module.

Where you have a resource that can be specified by a name or an ID I cannot find any sensible way in Terraform to give that resource a default name value apart from null. The problem is when you have the data sources like this

```
data “ibm_is_vpc” “vpc” {
 count = local.vpc_name != null ? 1 : 0
 name = local.vpc_name
}
```

You cannot introduce anything to `local.vpc` name that makes it not known when you run `terraform plan`. So for example adding this to your `locals {}`

```
 vpc_name = var.vpc_name == null && var.vpc_id == null ? “tok-vpc” : var.vpc_name
```

which is the logic I’d want to apply to set `local.vpc_name` to a default value, does not work in the e2e configuration because `var.vpc_id` comes from the output of another module and isn’t known until that module has finished executing, and so that ‘taints’ `local.vpc_name` and makes it not known at plan-time, and therefore unsuitable to use in the `count`.

To make running the Ansible playbooks work in the e2e Terraform you need to introduce a parameter for the Ansible playbook subdirectory. However, you cannot use variables in a destroy-time provisioner. So it’s a bit of hack but what I have to do is copy the ansible playbooks used in the destroy-time provisioner to the Terraform configuration’s base directory and run them from there.

**Comment by stevenhorsman** [comment]
Just a quick update on my review so far:
- I hit an issue with the `podvm_image_name` validation (and have chatted to Matt about it), but once I set a name that ended with `-amd64` I was able to get the e2e process running with a minimal set of variables supplied:
`ibmcloud_api_key`, `ibmcloud_user_id`,`cluster_name`,`ssh_key_name`,`podvm_image_name`,  `cos_bucket_name` and `ssh_pub_key`
I'm going to next try and create an s390x based cluster using a few more of the optional variables and then start testing the individual modules (the old/advanced approach).

**Comment by stevenhorsman** [comment]
I've now managed to get the e2e process tested with a s390x-based cluster, running in Toronto and specifying all the optional variables. I also tested the image architecture fix too.
Next up is to run through the individual module approach...

File: ibmcloud/README.md [hunk 1]
```diff
@@ -91,29 +188,24 @@ As usual, you need to create `terraform.tfvars` to specify parameter values. The
...
+primary_subnet_name = "<name of your primary subnet>" OR primary_subnet_id = "<ID of your primary subnet>"
+primary_security_group_name = "<name of your primary security group>" OR primary_security_group_id = "<ID of your primary security group>"
+vpc_name = "<name of your VPC>" OR vpc_id = "<ID of your VPC>"
 ```
+
 If you don't have your public key already configured in IBM Cloud you can add
 ```
 ssh_pub_key = "<your SSH public key>"
 ```
-> **Hint:** In order to create the cluster based on a different type of VSI image you can overwrite more parameters here e.g. to create a **s390x** based cluster add follow two lines to the `terraform.tfvars` file
```
Comment (stevenhorsman) [inline ibmcloud/README.md:99]: Optional/idea: I wonder if it is still worth repeating this section in the Cluster creation, or point to the section where it is mentioned in the e2e process to highlight that it's relevant at this stage?

File: ibmcloud/README.md [hunk 2]
```diff
@@ -91,29 +188,24 @@ As usual, you need to create `terraform.tfvars` to specify parameter values. The
 ibmcloud_api_key = "<your API key>"
 ssh_key_name = "<your SSH key name>"
 cluster_name = "<cluster name>"
+primary_subnet_name = "<name of your primary subnet>" OR primary_subnet_id = "<ID of your primary subnet>"
+primary_security_group_name = "<name of your primary security group>" OR primary_security_group_id = "<ID of your primary security group>"
+vpc_name = "<name of your VPC>" OR vpc_id = "<ID of your VPC>"
```
Comment (stevenhorsman) [inline ibmcloud/README.md:193]: Optional/idea: Should we provide the user more guidance of where to find these variables e.g. point out that the ids should be printed as output variables at the end of the **Create VPC** section?

File: ibmcloud/README.md [hunk 3]
```diff
@@ -171,14 +263,18 @@ https://cloud.ibm.com/objectstorage/
...
-cos_bucket_name = "<COS bucket name>"
-cos_service_instance_name = "<COS instance name>"
+cos_bucket_name = "<name of the COS bucket to create>"
+cos_service_instance_name = "<name of the COS instance to create>"
 ```
-> **Note:** The environment variable `cos_service_instance_name` in `variables.tf` has the default value `cos-image-instance` which will be used by Terraform if you do not provide a unique value in `terraform.tfvars`.
+
+> **Notes:**
+> - Variables with the same name as variables in the end to end Terraform configuration are as described in that [configuration's parameters](#parameters).
+> - Additional variables and their defaults are defined in the [variables.tf]()./terraform/cos/variables.tf) file for this Terraform configuration.
```
Comment (stevenhorsman) [inline ibmcloud/README.md:276]: It looks like the classic IDE auto-closing a bracket has happened here on `()./terraform/cos/variables.tf)`

File: ibmcloud/README.md [hunk 4]
```diff
@@ -273,31 +362,33 @@ $ install cloud-api-adaptor /usr/local/bin/
...
 ibmcloud_api_key = "<your API Key>"
 cluster_name = "<cluster name>"
-ssh_key_name = "<your SSH key name>"
-podvm_image_name = "<name of your pod VM image>"
+ssh_key_name = "<your SSH key name>" OR ssh_key_id = "<your SSH key id>"
+podvm_image_name = "<name of your pod VM image>" OR podvm_image_id = "<ID of your peer pod VM image>"
+vpc_name = "<name of your VPC>" OR vpc_id = "<ID of your VPC>"
+primary_subnet_name = "<name of your primary subnet>" OR primary_subnet_id = "<ID of your primary subnet>"
+primary_security_group_name = "<name of your primary security group>" OR primary_security_group_id = "<ID of your primary security group>"
 ```
```
Comment (stevenhorsman) [inline ibmcloud/README.md:377]: Optional/idea: I wonder if it is worth highlighting to the user that the `podvm_image_id` is provided as an output variable in the **Build a pod VM image** section and the `vpc_id`, `primary_subnet_id` and `primary_security_group_id` are output variables in the **Create a VPC section**?

File: ibmcloud/README.md [hunk 5]
```diff
@@ -314,18 +405,20 @@ After `terraform apply` completes the `cloud-api-adaptor` process will run on th
...
 ibmcloud_api_key = "<your API Key>"
 cluster_name = "<cluster name>"
+vpc_name = "<name of your VPC>" OR vpc_id = "<ID of your VPC>"
 ```
 
 > **Notes:**
-> - `ibmcloud_api_key` is your IBM Cloud API Key that you created at [https://cloud.ibm.com/iam/apikeys](https://cloud.ibm.com/iam/apikeys).
-> - `cluster_name` is a name of a Kubernetes cluster. You must use the same value for this parameter as you used for the corresponding parameter when running the Terraform template in [ibmcloud/terraform/cluster](terraform/cluster) to create the cluster.
+> - The `vpc` resource can be specified using its name or ID. The `vpc_name` and `vpc_id` variables are mutually exclusive.
```
Comment (stevenhorsman) [inline ibmcloud/README.md:419]: Optional/idea: Similar to the above comments about whether we want to point the user to the VPC output variables?

File: ibmcloud/README.md [hunk 6]
```diff
@@ -314,18 +405,20 @@ After `terraform apply` completes the `cloud-api-adaptor` process will run on th
...
-A Terraform template that will deploy an nginx pod to the Kubernetes cluster is available in [ibmcloud/terraform/run-nginx-demo](./terraform/run-nginx-demo).
+A Terraform configuration that will deploy an nginx pod to the Kubernetes cluster is available in [ibmcloud/terraform/run-nginx-demo](./terraform/run-nginx-demo). This configuration will also check the nginx peer pod virtual server instance has been successfully created.
 
-Create a `terraform.tfvars` file in the [template directory](./terraform/run-nginx-demo) for this Terraform template on your `development machine`. The `terraform.tfvars` file should look like this
+Create a `terraform.tfvars` file in the [configuration directory](./terraform/run-nginx-demo) for this Terraform configuration on your `development machine`. The `terraform.tfvars` file should look like this
 
 ```
 ibmcloud_api_key = "<your API Key>"
 cluster_name = "<cluster name>"
+vpc_name = "<name of your VPC>" OR vpc_id = "<ID of your VPC>"
```
Comment (stevenhorsman) [inline ibmcloud/README.md:415]: I think `podvm_image_id` , or `podvm_image_id` is a required field for this step as well?

**Review by stevenhorsman — COMMENTED** [review]
I've tested the individual steps and have a few doc comments, mostly minor/optional.

**Comment by liudalibj** [comment]
> @yoheiueda, @liudalibj - I've reviewed this from an end-user point of view, but my terraform skills are very limited, so if you are available to take a look from that perspective that would be great. Thanks!

I will take a look on this today.

**Review by liudalibj — COMMENTED** [review]
minimal string changes.
I run the E2E with default values, it cost about 45-50 mins.

**Review by yoheiueda — CHANGES_REQUESTED** [review]
Looks like the use of modules makes the test process much easier. Thank you very much!

File: ibmcloud/README.md [hunk 7]
```diff
@@ -188,34 +284,27 @@ $ terraform plan
...
+> - The `cos_service_instance` resource can be specified using its name or ID. The `cos_service_instance_name` and `cos_service_instance_id` variables are mutually exclusive.
+> - Variables with the same name as variables in the end to end Terraform configuration are as described in that [configuration's parameters](#parameters).
+> - All COS and VPC infrastructure resources must exist before running this Terraform configuration.
+> - Additional variables and their defaults are defined in the [variables.tf](./terraform/podvm-build/variables.tf) file for this Terraform configuration.
+> - If you don't specify the optional `podvm_image_name` variable then the name of the custom image created will be based on the latest commit hash of the [latest commit hash](https://github.com/confidential-containers/cloud-api-adaptor/commits/staging) of the confidential containers cloud API adaptor staging branch.
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: Thank you for spotting the repetition!

File: ibmcloud/README.md [hunk 8]
```diff
@@ -336,7 +429,7 @@ $ terraform plan
 $ terraform apply
 ```
 
-Deploying the demo workload will create a new nginx Pod and a NodePort service on your Kubernetes cluster, and a new Virtual Server instance for the peer pod will be created in your IBM Cloud VPC. The `run-nginx-demo` Terraform template will also sniff test the deployed nginx server by accessing the HTTP port of the NodePort service and test that the CPU architecture of the Kubernetes worker matches that of the peer pod instance.
+Deploying the demo workload will create a new nginx Pod and a NodePort service on your Kubernetes cluster, and a new virtual server instance for the peer pod will be created in your IBM Cloud VPC. The `run-nginx-demo` Terraform configuration will also sniff test the deployed nginx server by accessing the HTTP port of the NodePort service and test that the CPU architecture of the Kubernetes worker matches that of the peer pod instance.
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: Thanks. I noticed you added the config map feature to the demo. It's right to add that the demo will test that feature so I added it.

File: ibmcloud/README.md [hunk 9]
```diff
@@ -188,34 +284,27 @@ $ terraform plan
...
-Create the `terraform.tfvars` file in [the template directory](./terraform/podvm-build). The `terraform.tfvars` looks like this.
+Create the `terraform.tfvars` file in [the configuration directory](./terraform/podvm-build). The `terraform.tfvars` looks like this.
 ```
 ibmcloud_api_key = "<your API key>"
 ibmcloud_user_id = "<IBM Cloud User ID>"
 cluster_name = "<cluster name>"
-cos_service_instance_name = "<COS Service Instance Name>"
-cos_bucket_name = "<COS Bucket Name>"
+cos_service_instance_name = "<Name of your COS service instance>" OR cos_service_instance_id = "<ID of your COS service instance>"
+cos_bucket_name = "<COS bucket name>"
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: I made that style change

File: ibmcloud/README.md [hunk 10]
```diff
@@ -171,14 +263,18 @@ https://cloud.ibm.com/objectstorage/
 
 First, create a COS service instance if you have not create one. Then, create a COS bucket with the COS instance. The COS service instance and bucket names are necessary to upload a custom VM image.
 
-You can use the Terraform template located at [ibmcloud/terraform/cos](./terraform/cos)to use Terraform to create a COS service instance, COS bucket, and IAM AuthorizationPolicy automatically. These resources are configured to store images. Create a `terraform.tfvars` file in the templates directory that includes these fields:
+You can use the Terraform configuration located at [ibmcloud/terraform/cos](./terraform/cos) to use Terraform to create a COS service instance, COS bucket, and IAM Authorization Policy automatically. These resources are configured to store images. Create a `terraform.tfvars` file in the configurations directory that includes these fields:
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: Agreed. I made that change.

File: ibmcloud/README.md [hunk 11]
```diff
@@ -53,35 +53,132 @@ Optionally, you can install IBM Cloud CLI.
...
 * Floating IP for the public gateway
+* Virtual server instances for the Kubernetes control plane and worker
+* Floating IPs for the Kubernetes control plane and worker virtual server instances
+* COS Instance with 1 COS bucket
+* Custom image for the peer pod VM image
+* Virtual server instance for the peer pod running the nginx workload
+* SSH key, if you specified the optional `ssh_pub_key` variable
+
+On a cluster using `instance_profile_name = "bz2-2x8"` and `image_name = "ibm-ubuntu-18-04-1-minimal-s390x-3"` the end-to-end playbook takes approximately 50 minutes to complete. 
+
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md:157]: Yes. I added an extra header

## Create and test the demo environment on IBM Cloud for peer pod VMs - advanced usage

File: ibmcloud/README.md [hunk 12]
```diff
@@ -53,35 +53,132 @@ Optionally, you can install IBM Cloud CLI.
...
 ```bash
-$ cd ibmcloud/terraform/common
+$ cd ibmcloud/terraform
 $ terraform init
 $ terraform plan
 $ terraform apply
 ```
 
-The following cloud resources will be created. Please check [main.tf](terraform/common/main.tf) for the details.
+The following IBM Cloud resources will be created. Please check [main.tf](terraform/common/main.tf) for the details.
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: Thanks for spotting this. I reworded that line so we can see if it's clearer now.

File: ibmcloud/README.md [hunk 13]
```diff
@@ -53,35 +53,132 @@ Optionally, you can install IBM Cloud CLI.
...
+
+You can create the demo environment for peer pod VMs on IBM Cloud Virtual Private Cloud (VPC) with the Terraform configuration located in [ibmcloud/terraform](./terraform). This Terraform configuration will:
+
+* Set up VPC infrastructure network resources, including the VPC, Subnets and Security groups
+* Set up VPC infrastructure compute resources, including virtual server instances for the Kubernetes control plane and worker
+* Optionally import a SSH public key to VPC Infrastructure
+* Build and install dependencies on and configure the Kubernetes control plane and worker instances
+* Build the peer pod VSI on the Kubernetes worker instance
+* Set up IBM Cloud Object Storage (COS) resources
+* Push the built peer pod VSI to COS and import it from COS to VPC Infrastructure as a Custom image
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: Agreed. I made that change.

File: ibmcloud/README.md [hunk 14]
```diff
@@ -53,35 +53,132 @@ Optionally, you can install IBM Cloud CLI.
...
 
-First, you need to create a Virtual Private Cloud (VPC). The Terraform configuration files are in [ibmcloud/terraform/common](./terraform/common/).
+### End to end Terraform configuration
+
+You can create the demo environment for peer pod VMs on IBM Cloud Virtual Private Cloud (VPC) with the Terraform configuration located in [ibmcloud/terraform](./terraform). This Terraform configuration will:
+
+* Set up VPC infrastructure network resources, including the VPC, Subnets and Security groups
+* Set up VPC infrastructure compute resources, including virtual server instances for the Kubernetes control plane and worker
+* Optionally import a SSH public key to VPC Infrastructure
+* Build and install dependencies on and configure the Kubernetes control plane and worker instances
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: Thanks for pointing out that this point doesn't read well. I split the one point into two points so it should read better now.

File: ibmcloud/README.md [hunk 15]
```diff
@@ -53,35 +53,132 @@ Optionally, you can install IBM Cloud CLI.
...
+* Build the peer pod VSI on the Kubernetes worker instance
```
[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply ibmcloud/README.md]: Agreed. I made that change.

File: ibmcloud/README.md (hunk 3 above)
Comment (mattarnoatibm) [reply ibmcloud/README.md]: I suspect you're exactly right. Good spot there. I removed the extra right bracket.

File: ibmcloud/README.md (hunk 1 above)
Comment (mattarnoatibm) [reply ibmcloud/README.md:99]: I've added the hint back.

File: ibmcloud/README.md (hunk 2 above)
Comment (mattarnoatibm) [reply ibmcloud/README.md:196]: I added a section under this part saying if the input variables can come from another Terraform configuration where they come from and where to find them, if the user used the Terraform configurations to create the resources (the user might have done it manually).

I did similar sections for the `podvm-build`, `start-cloud-api-adaptor` and `run-nginx-demo` Terraform configurations as well, so these sections should read similarly.

File: ibmcloud/README.md (hunk 4 above)
Comment (mattarnoatibm) [reply ibmcloud/README.md:386]: I did the same thing here as I did in the https://github.com/confidential-containers/cloud-api-adaptor/pull/49#discussion_r893355021 comment

File: ibmcloud/README.md (hunk 5 above)
Comment (mattarnoatibm) [reply ibmcloud/README.md:433]: I did the same thing here as I did in the https://github.com/confidential-containers/cloud-api-adaptor/pull/49#discussion_r893355021 comment.

**Comment by mattarnoatibm** [comment]
Thanks for all the feedback, Yohei @yoheiueda , Da Li @liudalibj and Steve @stevenhorsman.

I should have addressed it in my latest commit so this PR should be ready to be re-reviewed

**Review by stevenhorsman — APPROVED** [review]
Thanks for all the changes. LGTM!

**Review by yoheiueda — APPROVED** [review]
LGTM

**Pull request merged**